### ISSUE TITLE
unable to use rezised images through public url

### ISSUE THREAD

**Issue by harshapulikollu** [firebase/extensions#140]
### [READ] Step 1: Are you in the right place?
Yes.

### [REQUIRED] Step 2: Describe your configuration

- Extension name: **\_** (`storage-resize-images`)
- Extension version: **\_0.1.3**

### [REQUIRED] Step 3: I'm unable to access the resized images through a public(download) url from my flutter app, like I'm able to retrieve original image. I don't see any reason to use this extension if i can't access those resized images by manipulating the original Image URL.
please provide me a way how can I access those resized images from my flutter app.

**Comment by brandonpearcy** [comment]
+1 to this. I have not found a convenient way to make the resized images publicly accessible without doing manual post-processing with additional cloud functions. 

The generated thumbs have no access token visible in Firebase console, there is no clickable link to view the images, and the preview of the image never loads (infinite spinner).

<img width="1267" alt="Screen Shot 2019-12-21 at 10 57 12 AM" src="https://user-images.githubusercontent.com/17098317/71312534-f8adaf00-23e0-11ea-9b29-e0936cfc026b.png">

I would expect that either: 

(1) there exists a simple transformation from the original firebase storage URL to access the resized images (e.g., something you could hardcode in client logic), or 
(2) this firebase extension comes with documentation describing how to best post-process the resized images to make them accessible

**Comment by NicolasMaas** [comment]
> +1 to this. I have not found a convenient way to make the resized images publicly accessible without doing manual post-processing with additional cloud functions.
> 
> The generated thumbs have no access token visible in Firebase console, there is no clickable link to view the images, and the preview of the image never loads (infinite spinner).
> 

I'm facing the same issue... Seems kind of pointless at the moment to use this extension at all.

**Comment by TiagoGouvea** [comment]
Happening with me also.

On my "dev" project, sending a new image trough firebase console or by code, it generates the public URL of the resized file.
- [Resized image on dev project](https://firebasestorage.googleapis.com/v0/b/whereami-dev.appspot.com/o/images%2F98453cef-5e3a-4423-93ce-7fde2e61c67b_650x500.jpg?alt=media&token=5ff48935-88dc-4973-9968-5c13001a9408)

On my "prod" project it don't.
-  [Original file on prod project](https://firebasestorage.googleapis.com/v0/b/guess-where-i-am.appspot.com/o/images%2F4433b4d7-f058-4aae-8faf-ffd11c42a4dd.jpg?alt=media&token=1285e547-0b18-4599-a3a7-ce2b392eb494)
- Resized filename: 4433b4d7-f058-4aae-8faf-ffd11c42a4dd_650x500.jpg
- Storage file: gs://guess-where-i-am.appspot.com/images/4433b4d7-f058-4aae-8faf-ffd11c42a4dd_650x500.jpg

A console print with infinite loading:

![Where-Am-I_–_Firebase_console](https://user-images.githubusercontent.com/2242549/71671691-73b48500-2d52-11ea-9767-e3b101d392b6.png)

Both projects has billing activated

**Comment by smbowler** [comment]
Just FYI, I'm following along on this thread. I'm hoping a solution is found soon. Thanks

**Comment by smbowler** [comment]
Hi @Ehesp, happy New Year. I just wanted to check-in to see if there was a solution to this issue. It seems to be the same issue presented in #155 

My development team paused development on our own solution when we found the firebase photo-resizing extension, but are awaiting your response to these two issues before we dedicate time to implementing and debugging. 

Any insight would be greatly appreciated. Thank you!

**Comment by lars-sorensen** [comment]
I am also wondering this. Also wondering about the lack response here...

**Comment by anthowm** [comment]
In my backend I generate signedUrls dont like this way but also is weird that in my frontend I need to delay the get of resized images because it return to me 404 if not delay it for 2000 (the next times to access to resized images works without delay...). 
When the original file works like a charm with signedUrl ...

First call is original file the second is the resized:
![image](https://user-images.githubusercontent.com/5764698/72948524-b5e74a00-3d85-11ea-9b56-f442e4572453.png)

```
for (const resizeFile of urisArray) {
          const bucketFile = this.storage.file(resizeFile);
          const result = await bucketFile.getSignedUrl({
            action: 'read',
            expires: '01-01-2050',
          });
          resizedImages.push(result[0] ? result[0] : '');
        }
```

**Comment by NickFoden** [comment]
In the extension configuration steps within the firebase console, there needs to be an option to enable the access token for the created thumbnails or could/should even be defaulted to a valid access token similar to behavior of the regular file upload to storage. 

This is crucial for being able to use this extension. Was searching on how to use the created thumbnails and have only hit dead ends. I am not sure which use case is currently being solved with this extension if the generated thumbnails are not available for use? Seems we have to go another direction for thumbnail generation for now. If anyone has any ideas how to access the thumbnails please ping me. Anyhow looking forward to future version of this image resize extension.  Txs all.

**Comment by lsps9150414** [comment]
@NickFoden Have been stuck with this as well. My workflow is:
Client upload to storage => resize extension => onStorageFinalize => cloud function to blur image & upload to storage again

So in the end I have:
- the original uploaded image
- the resized images
- the blurred images

I cannot find a solution to generate/get the downloadable URLs (containing the tokens) for the resized & blurred images in the context of my cloud function in order to store to my Firestore DB so I ended up with only storing the file path (the `name` prop. from the data object) in the DB and doing `getDownloadURL`  on the client-side, which is obviously not ideal tho but is the only way I have found so far.

**Comment by DhavalW** [comment]
+1 & I don't think this is a token access issue.

In my workflow, I'm serving some images & assets from a path fully open to public. ie Access rules are set to :
```
match /public/{allPaths=**}{
    	allow read, write;
    }
```
For every asset I manually construct the public url for this path, and purposely do not include any access tokens. So far all images were served just fine. However images generated by this extension are not served.

Unfortunately that's a deal killer for me,  so i will have uninstall this extension & write my own image resizer. **dread**

**Comment by NickFoden** [comment]
If you go find the new asset in your storage you can see the issue. Say you upload random.png go view in the web console and find it in storage. Expand file location and see an Access token
![image](https://user-images.githubusercontent.com/8321838/75581746-1e39e300-5a38-11ea-8194-91d4ea56b8ce.png)

Now do the same thing for the generated thumbnail say random_200x200.png. There is no access token set, and so you will get a 403 trying to use it in your application unless you go enable it's access token manually. But no way I can do this for every single crop being uploaded. This was my experience a few weeks back trying this out. Will need to have the access token set up when the cropped image is created, same as a normal image uploaded gets a working access token.

**Comment by DhavalW** [comment]
@NickFoden Yep. tried that just afterwards and you're right - it works once the token is enabled. & since its manual, doing so at scale is simply not an option. 

But why does it occur even when auth requirements are disabled in rules ? that doesn't make sense. All other images on that path get served just fine without a token.

Also shouldn't this issue be marked as a bug ?

**Comment by Ehesp** [comment]
Implementation/fix details are being discussed. Once in place, I'll update this issue and start working on it... Thanks for your patience!

**Comment by dfdgsdfg** [comment]
signedUrl error log with non exist token
I create signedUrl with adminStorage.bucket(bucketName).file(filePath).getSignedUrl(options), but it does not work.

<Error>
<Code>SignatureDoesNotMatch</Code>
<Message>
The request signature we calculated does not match the signature you provided. Check your Google secret key and signing method.
</Message>
<StringToSign>
GET 1895632947 /project.appspot.com/users/NK8bDKYlaofgwkX6cxy14tqpsGa2/room_backgrounds/room_bg_1584592936559_Ib6R9sawuIQ_200x200.jpeg
</StringToSign>
</Error>
If I create access token on web ui after created signedUrl, it still does not work.

**Comment by QuentinSc** [comment]
Hello,

No news ? :)

**Comment by DhavalW** [comment]
+1 waiting for a fix

**Comment by K0EZ** [comment]
this fix worked for me 
https://stackoverflow.com/questions/59072882/cloud-functions-resized-images-not-loading/61023298#61023298

**Comment by sitefinitysteve** [comment]
How is this not fixed yet, I thought this extension solved all my avatar issues.  But without the tokens it's useless.

The solution from @K0EZ worries me if the function code gets somehow gets updated by Firebase on a new version?

Feels like a no brainer to just make this a config option in the extension options...?

**Comment by sitefinitysteve** [comment]
Submitted a PR to block the delete based on a config option, no idea if it'll get approved though...

**Comment by laurenzlong** [comment]
Thanks for the contribution @sitefinitysteve, the reason why we took so long was because we didn't realize that the issue was simply that we should not be deleting the download token from the metadata. Thanks for providing the Stackoverflow link @K0EZ !

**Comment by NickFoden** [comment]
Yeah the PR is especially considerate, to even ask the user if they want to delete the token. But pretty sure can just remove that one line

`delete metadata.metadata.firebaseStorageDownloadTokens;`

And everyone can finally use this extension. Unless maybe someone does need to hide their thumbnails, but not the originals? (Seems like that would be a rare use case.)

**Comment by Mian-Zaid** [comment]
I have found the solution to it like i have to comment out the line which deleting the access token in the cloud function and now i have access to token of image after resizing but there is problem that if i have to change the configurations of extension the code is updated in cloud function so i have to change this line every time i change the configuration.
this solution works for me well but its not a solid solution.

its simple process

1. after installing the extension go to functions
2. then select **view details usage** its opens a screen from where you have to select **source**
![vd](https://user-images.githubusercontent.com/38668413/80060773-cf488f80-8548-11ea-86c7-ea2a01d0a24a.PNG)
3. form source find file named **FUNCTIONS/LIB/INDEX.JS** , Go to line 152 then comment this out
 code on line 152 : **delete metadata.metadata.firebaseStorageDownloadTokens;**
![file](https://user-images.githubusercontent.com/38668413/80060782-d7083400-8548-11ea-9d7c-7d3777a427c7.PNG)

**Comment by karayu** [comment]
Thanks everyone! I think we pushed a commit last week that fixed this. Please let us know if you continue to see issues!

**Comment by devth** [comment]
I just installed this extension, and by default it looks like the resized images don't have tokens, so I can't see them in the Firebase Console. Is that expected?

**Comment by felix-lambert** [comment]
Yes same for me I don't have the public token url

**Comment by dackers86** [comment]
Hi @devth @felix-lambert 

I'll reopen this issue for investigation.

**Comment by dackers86** [comment]
Hi @devth @felix-lambert.

I am currently unable to replicate a missing token, can you confirm which version of the extension you are using?

The latest version is currently `storage-resize-images@0.1.22`

**Comment by felix-lambert** [comment]
Hello thanks for your help. Yes I can confirm that the public token url is not available on the resized image in version 0.1.22. I only have the storage (gs://....png) but not the access token. The image with the original size has the access token but not the resized image.

**Comment by dackers86** [comment]
Can you confirm that the `File location` tab on the right does not contain a token in Firebase storage?

Apologies if I am looking in the incorrect location.

**Comment by felix-lambert** [comment]
I have a file location tab. It contains: gs://.../mint.png for example  but I don't have an Access token tab on the right for the resized image.

**Comment by i14h** [comment]
@dackers86 lets reevaluate where we are on this

**Comment by dackers86** [comment]
Possibly related https://github.com/firebase/extensions/issues/998

**Comment by yamankatby** [comment]
Happy to close this issue as `completed` as this feature is available now starting from `v0.1.29` we have a config option to make images public that can be toggled during installation or by updating the extension config.

![Screen Shot 2022-09-07 at 12 56 16](https://user-images.githubusercontent.com/35961879/188849842-5e5905fe-c3e2-4068-809d-22fd453dde40.png)

Feel free to reopen it if you still have issues with it

**Comment by CalvinJamesHeath** [comment]
@yamankatby I have it set to false but the images are still public. I'm I missing something?

**Comment by jauntybrain** [comment]
Hi @CalvinJamesHeath! Could you please provide more information about this issue? Most importantly, what URL are you using to access the image?

**Comment by CalvinJamesHeath** [comment]
@jauntybrain
Example URL of resized image with extension set to false public images : [](https://firebasestorage.googleapis.com/v0/b/gbapp-b859e.appspot.com/o/postImages%2F5f79c237-a7d5-4c72-b0ae-6dcda1cd625b_p54FytR8xeVAcmliRJ3GqOxCUeZ2_1080x1080.jpeg?alt=media&token=16d6e8fa-63ce-4764-ac59-bfde0abaa306)

**Comment by CalvinJamesHeath** [comment]
The [documentation](https://firebase.google.com/docs/storage/security/secure-files#understand_rules) explains that if no rule expression is provided, the rule evaluates to true. I see now. Thank you.
